### PR TITLE
GHC Core

### PR DESCRIPTION
**Pull request by supersven**
Add Lexer for GHC Core

Core is an intermediate language of the Glasgow Haskell Compiler.

### CODE REVIEW

File: lib/rouge/lexers/ghc_core.rb [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      state :name do
+        rule %r/^([A-Z]\w*)(\.)/ do |m|
+          token Name::Namespace, m[0]
+          token Punctuation, m[1]
+        end
+
+        rule %r/[A-Z][^\s.,(){}]*/, Keyword::Type
+
+        # packages, e.g. 'ghc-prim-0.5.3:'
+        rule %r/(^[a-z].*?\d+\.\d+\.\d+)(:)(?=\S+)/ do |m|
```
Comment (pyrmont) [inline lib/rouge/lexers/ghc_core.rb]: Would having the positive lookahead be `(?=\S)` not be sufficient? It seems wasteful to be matching multiple characters for a lookahead.

File: lib/rouge/demos/ghc-core [hunk 2]
```diff
@@ -0,0 +1,55 @@
+-- RHS size: {terms: 23, types: 50, coercions: 0, joins: 0/0}
```
[the remark this answers is not in the document]
Comment (supersven) [reply lib/rouge/demos/ghc-core]: Hey @pyrmont ,
Thanks for the fast feedback! 😄 

It was real fun to work with such a well-crafted lexer framework.

I'll add a smaller demo. It was useful to have a big real-world example during development, but I understand that it's really too big.

File: lib/rouge/lexers/ghc_core.rb [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        mixin :common
+
+        rule %r/\n/, Text, :pop!
+      end
+
+      state :annotation do
+        rule %r/\]/, Comment::Special, :pop!
+        rule %r/\[/, Comment::Special, :annotation
+        rule %r/[^\[\]]+/, Comment::Special
+        rule %r/\s+/, Comment::Special
```
[the remark this answers is not in the document]
Comment (supersven) [reply lib/rouge/lexers/ghc_core.rb]: Good spot! Thanks.

File: lib/rouge/demos/ghc-core (hunk 2 above)
[the remark this answers is not in the document]
Comment (supersven) [reply lib/rouge/demos/ghc-core]: Sorry, I mixed up `demos` and `samples` ...

Is the size of the `demo` now okay?

Can the `sample` stay as it is or should I choose a smaller one?

[the remark this answers is not in the document]
Comment (supersven) [reply lib/rouge/demos/ghc-core]: Hm, now I cut down `samples`, too :)

[the remark this answers is not in the document]
Comment (pyrmont) [reply lib/rouge/demos/ghc-core]: Looks good!

**Comment by pyrmont** [comment]
@supersven Sorry, one minor change: I renamed the class to be a little more consistent with the general approach taken with Rouge's lexers. It's not always universally followed but it should be done this way.

**Comment by supersven** [comment]
Hey @pyrmont ,

Thanks a lot for the review and your changes!

I forgot to mention one thing: If you like, you can squash the commits during merge. I think that my "baby step" commits might be more confusing than helping in the long run ... 😄

**Comment by pyrmont** [comment]
@supersven Yep, I always squash the PR commits into a single commit before merging. Helps to keep the commit history nice and clean :) Thanks again for all the work! 🎉

The next version of Rouge will be out on Tuesday 14 January. This will be part of that release :)

**Comment by supersven** [comment]
@pyrmont ,
Fantastic! Thanks a lot for the review! 👍

**Pull request merged**